src/time.rs

```diff
@@ -0,0 +1,281 @@
+
+#[cfg(feature = "std")]
+const NANOS_PER_MILLI: u32 = 1_000_000;
+
+/// A representation of an absolute time value.
+///
+/// The `Instant` type is a wrapper around a `i64` value that
+/// represents a number of milliseconds, monotonically increasing
+/// since an arbitrary moment in time, such as system startup.
+///
+/// * A value of `0` is inherently arbitrary.
+/// * A value less than `0` indicates a time before the starting
+///   point.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Instant {
+    millis: i64,
+}
+
+impl Instant {
+    /// Create a new `Instant` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Instant {
+        Instant { millis }
+    }
+
+    /// The fractional number of milliseconds that have passed
+    /// since the beginning of time.
+    pub fn millis(&self) -> i64 {
+        self.millis % SIGNED_MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds that have passed since the
+    /// beginning of time.
+    pub fn secs(&self) -> i64 {
+        self.millis / SIGNED_MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds that have passed since
+    /// the biginning of time.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Instant {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis + rhs.total_millis() as i64)
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis - rhs.total_millis() as i64)
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: u64,
+}
+
+impl Duration {
+    /// Create a new `Duration` from a number of milliseconds.
+    pub fn from_millis(millis: u64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The fractional number of milliseconds in this `Duration`.
+    pub fn millis(&self) -> u64 {
+        self.millis % UNSIGNED_MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds in this `Duration`.
+    pub fn secs(&self) -> u64 {
+        self.millis / UNSIGNED_MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds in this `Duration`.
+    pub fn total_millis(&self) -> u64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Duration {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Duration;
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration::from_millis(
+            self.millis.checked_add(rhs.total_millis()).expect("overflow when adding durations"))
```

Comment:
I maintain my position that checking for unsigned overflow is a waste of time.

Reply:
Np, I'll revert it. Just wondered if that was enough to used the `checked` variant.